### ISSUE TITLE
Convert remaning runtime helpers to TypeScript

### ISSUE THREAD

**Issue by nicolo-ribaudo** [babel/babel#16500]
Our source files are mostly written in TypeScript, with the exception of most of the runtime helpers that we inject in the compiled code. Since https://github.com/babel/babel/pull/16495 all those helpers are stored in individual files and pre-transpiled, so we can move all those that are still written in JS to TypeScript.

There are some rules/instructions to follows:
- Many imports currently use "bare specifiers", such as `import toPropertyKey from "toPropertyKey";`. They must be converted to relative imports **always using the `.ts` extension**, such as `import toPropertyKey from "./toPropertyKey.ts";`. The reason for using `.ts` even when importing `.js` files is that, when this issue will be completed, all those files will be using `.ts` anyway. If TS complain that it cannot find the types of the dependency because it's still written in JS, add `// @ts-expect-error` before the import.
- The files must type-heck when running `make tscheck-helpers` (or `yarn tsc -p ./packages/babel-helpers/src/helpers/tsconfig.json`, which is equivalent)
- Some of those helpers use absolutely unreadable variable names, such as https://github.com/babel/babel/blob/db3e9a6cbdaba16dc480a1f76a11b0d2ea84adc4/packages/babel-helpers/src/helpers/iterableToArrayLimit.js, because those files were "manually minified". No we minify helpers at build time, so we can afford to use readable variables: feel free to change them!

To coordinate this work, I split all our helpers in groups: those who want to help can leave a comment saying which group you want to work, and _then_ start working on a PR :) If you want to work on multiple groups, start by claiming **one** and then after that the PR is merged you can claim a second one (and so on).

If it is the first time that you contribute to Babel, follow these steps: (you need to have `make` and `yarn` available on your machine)
1. Fork the repo
1. Run `git clone https://github.com/<YOUR_USERNAME>/babel.git && cd babel`
1. Run `yarn && make bootstrap`
1. Wait :hourglass_flowing_sand: 
1. Update the code!
1. Whenever you change a helper, you'll need to run `yarn gulp generate-runtime-helpers` to regenerate `./packages/babel-helpers/src/helpers-generated.ts` (this is probably not needed if you _just_ add types, but it is if you for example change variable names)
1. Once you are done, you can run `make build` to re-build everything, and then `make test` to run all the tests
1. Run `git push` and open a PR!

---

**NOTE**: Some of the helpers in this list are already written in TS, but they are failing type-checking. The task for those is to remove them from `packages/babel-helpers/src/helpers/tsconfig.json`'s `exclude` list and fix the type errors.

- [x] **Legacy decorators** (@amjed-98)
  - `applyDecoratedDescriptors`
  - `initializerDefineProperty`
  - `initializerWarningHelper`
- [x] **New decorators** (@liuxingbaoyu)
  - `defineAccessor`
  - `identity`
  - `setFunctionName` (see note above)
  - `applyDecs2311` (see note above)
- [x] **Arrays** (@blakewilson)
  - `arrayLikeToArray`
  - `arrayWithHoles`
  - `arrayWithoutHoles`
  - `iterableToArray`
  - `iterableToArrayLimit`
  - `iterableToArrayLoose`
  - `maybeArrayLike`
  - `unsupportedIterableToArray`
  - `slicedToArray`
  - `slicedToArrayLoose`
  - `toArray`
  - `toConsumableArray`
- [x] **Objects** (@SukkaW)
  - `defaults`
  - `extends`
  - `objectDestructuringEmpty`
  - `objectSpread2`
  - `objectWithoutProperties`
  - `objectWithoutPropertiesLoose`
- [x] **Reflection** (@coelhucas)
  - `defineEnumerableProperties`
  - `defineProperty`
  - `get`
  - `getPrototypeOf`
  - `set`
  - `setPrototypeOf`
  - `superPropBase`
  - `construct` (see note above)
- [x] **Classes** (@amjed-98)
  - `assertThisInitialized`
  - `classCallCheck`
  - `createClass`
  - `possibleConstructorReturn`
- [x] **Subclasses** (@amjed-98)
  - `callSuper`
  - `inherits`
  - `inheritsLoose`
  - `isNativeFunction`
  - `wrapNativeSuper`
- [x] **For-of** (@SukkaW)
  - `creaeForOfIteratorHelper`
  - `creaeForOfIteratorHelperLoose`
- [x] **Async functions/generators** (@SukkaW)
  - `asyncToGenerator`
  - `AsyncGenerator`
  - `asyncGeneratorDelegate`
  - `asyncIterator`
  - `awaitAsyncGenerator`
  - `OverloadYield`
  - `wrapAsyncGenerator`
- [x] **Class fields** (@Zzzen)
  - `checkPrivateRedeclaration`
  - `classPrivateFieldInitSpec`
  - `classPrivateMethodInitSpec`
  - `classStaticPrivateMethodGet`
  - `classPrivateFieldLooseBase`
  - `classPrivateFieldLooseKey`
  - `readOnlyError`
  - `writeOnlyError`
- [x] **Modules** (@amjed-98)
  - `interopRequireDefault`
  - `interopRequireWildcard`
  - `importDeferProxy`
- [x] **Temporal dead zone** (@amjed-98)
  - `tdz`
  - `temporalRef`
  - `temporalUndefined`
- [x] **Other** (@SukkaW)
  - `jsx`
  - `taggedTemplateLiteral`
  - `taggedTemplateLiteralLoose`
  - `instanceof`
  - `typeof`
  - `wrapRegExp`
  - `usingCtx` (see note above)

**Comment by nicolo-ribaudo** [comment]
Thank you! If you need any help feel free to ask :)

**Comment by SukkaW** [comment]
I am going to work on `Objects` first!

**Comment by nicolo-ribaudo** [comment]
Thanks!

**Comment by coelhucas** [comment]
I can get `Reflection` 🙂

**Comment by Zzzen** [comment]
`Class fields` please!

**Comment by nicolo-ribaudo** [comment]
They are yours :)

@coelhucas After that https://github.com/babel/babel/pull/16505 is merged, you might need to rebase your changes to remove some `@ts-expected-error` comments introduced by that PR.

**Comment by arslivinski** [comment]
I would like to work on the `For-of`.

**Comment by nicolo-ribaudo** [comment]
It's yours :)

**Comment by SukkaW** [comment]
After taking a closer look, I want to also give **Async functions/generators** a shot.

**Comment by magic-akari** [comment]
I am concerned about using TypeScript with Babel helpers. Does type really make sense?

In my experience, Babel helper is already js in downstream users' node_modules, and usually our tsc also turns on `skipLibCheck`.
The only place where it can make a difference is in the Babel repository itself for checking.

Do we have enough TypeScript code using Babel helpers so that the helper is imported and checked by `tsc`?
If the function is exported but not used, in such a case, Type has not fulfilled its purpose at all.

**Comment by nicolo-ribaudo** [comment]
@magic-akari Yes, helpers being written in TS is mostly for internal type checking in the monorepo (among the helpers themselves). All the helpers we added in the past few years are already written in TS because:
- it makes the DX when working on them much better, as autocompletion works better
- helpers can be imported by other helpers (even if many of them are not), exercising the cross-helper checks
- helpers are written in a very weird JavaScript, to maximize compatibility and minimize code size, and having types helps a lot with documenting what is going on

When published, helpers are never meant to be manually imported in code that people write, but they are just injected by Babel and thus at that point don't need type checking.

**Comment by amjed-98** [comment]
can I have `Classes` ?

**Comment by nicolo-ribaudo** [comment]
Yes!

**Comment by nicolo-ribaudo** [comment]
Yes!

**Comment by amjed-98** [comment]
`Modules`?

**Comment by amjed-98** [comment]
quick question
can you give me a usage example for the `_identity` helper function in `packages/babel-helpers/src/helpers/identity.ts`?

@nicolo-ribaudo

**Comment by nicolo-ribaudo** [comment]
No constraint. I think currently we only pass functions to it, but let's keep it as generic as possible since we might use for other things.

**Comment by amjed-98** [comment]
`legacy decorators?`

**Comment by SukkaW** [comment]
I will tackle `Other` next.

**Comment by amjed-98** [comment]
can I have `Subclasses` next?
@nicolo-ribaudo

**Comment by mo7amedehab97** [comment]
can i work on Subclasses => callsuper @nicolo-ribaudo

**Comment by arslivinski** [comment]
I'm not having spare time to work on the `for-of`, so if anyone wants to do it, feel free.

**Comment by SukkaW** [comment]
And I am gonna work on `For-of` next.

**Comment by SukkaW** [comment]
@nicolo-ribaudo Just a friendly ping as the last pieces of the issue have been finished. After merging #16578 and #16585 the issue could be closed.
